cpp/src/lookup/lookup_src_dst_impl.cuh

```diff
@@ -0,0 +1,1058 @@
+                             gpu_offsets.begin(),
+                             gpu_offsets.end(),
+                             gpu_offsets.begin(),
+                             size_t{0});
+
+      h_unique_gpu_ids.resize(unique_gpu_ids.size());
+      h_gpu_offsets.resize(gpu_offsets.size());
+      raft::update_host(
+        h_unique_gpu_ids.data(), unique_gpu_ids.data(), unique_gpu_ids.size(), handle.get_stream());
+      raft::update_host(
+        h_gpu_offsets.data(), gpu_offsets.data(), gpu_offsets.size(), handle.get_stream());
+
+    } else {
+      thrust::sort(handle.get_thrust_policy(), edgelist_types.begin(), edgelist_types.end());
+
+      h_unique_gpu_ids.resize(size_t{1});
+      h_unique_gpu_ids[0] = 0;
+
+      h_gpu_offsets.resize(h_unique_gpu_ids.size() + 1);
+      h_gpu_offsets[0] = 0;
+      h_gpu_offsets[1] = edgelist_types.size();
+    }
+
+    //
+    // For edge ids mapped to each gpu, count number of unique types and elements per type.
+    // cub::DeviceSegmentedReduce(ByKey) ???
+    //
+
+    [[maybe_unused]] std::vector<int> h_gpu_ids_partition{};
+    std::vector<edge_type_t> h_types_partition{};
+    std::vector<edge_t> h_type_freqs_partition{};
+
+    rmm::device_uvector<edge_type_t> unique_types_segment(0, handle.get_stream());
+    rmm::device_uvector<edge_t> type_freqs_segment(0, handle.get_stream());
+
+    for (size_t idx = 0; idx < h_unique_gpu_ids.size(); ++idx) {
+      auto gpu_id = h_unique_gpu_ids[idx];
+
+      auto nr_uniqe_types_segment = thrust::count_if(
+        handle.get_thrust_policy(),
+        thrust::make_counting_iterator(size_t{0}),
+        thrust::make_counting_iterator(
+          static_cast<size_t>(h_gpu_offsets[idx + 1] - h_gpu_offsets[idx])),
+        detail::is_first_in_run_t<edge_type_t const*>{edgelist_types.data() + h_gpu_offsets[idx]});
+
+      unique_types_segment.resize(nr_uniqe_types_segment, handle.get_stream());
+      type_freqs_segment.resize(nr_uniqe_types_segment, handle.get_stream());
+
+      thrust::reduce_by_key(handle.get_thrust_policy(),
+                            edgelist_types.begin() + h_gpu_offsets[idx],
+                            edgelist_types.begin() + h_gpu_offsets[idx + 1],
+                            thrust::make_constant_iterator(size_t{1}),
+                            unique_types_segment.begin(),
+                            type_freqs_segment.begin());
+
+      std::vector<edge_type_t> h_unique_types_segment(nr_uniqe_types_segment);
+      std::vector<edge_t> h_type_freqs_segment(nr_uniqe_types_segment);
+
+      raft::update_host(h_unique_types_segment.data(),
+                        unique_types_segment.data(),
+                        unique_types_segment.size(),
+                        handle.get_stream());
+      raft::update_host(h_type_freqs_segment.data(),
+                        type_freqs_segment.data(),
+                        type_freqs_segment.size(),
+                        handle.get_stream());
```

Comment:
Don't forget sync_stream() after update_host. Data might be invalid when you call insert below.

Reply:
Added.